### DISCUSSION TITLE
How can I detect in my init.vim whether I use Neovide or not?

### DISCUSSION THREAD

**Discussion by unclechu** [neovide/neovide#919] [Q&A] [answered]
Different GUIs have different ways of setting font family & size for instance and I usually detect what GUI am I using in the config to apply one setting or another. Is there any way to detect synchronously whether it is Neovide or not?

Here how I detect whether it is Neovim-Gtk or not:

``` viml
if !exists('s:is_neovim_gtk_gui')
	let s:is_neovim_gtk_gui = exists('g:GtkGuiLoaded') ? 1 : 0
en
```

Neovim-Gtk sets `g:GtkGuiLoaded` synchronously. I only found out that Neovide sets some global variables after start up but it happens very late. I wrote this dirty hack to detect it:

``` viml
fu! g:MyGuiInitCallback(timer)
	if exists('*g:NeovideNotifyfullscreenChanged')
		ec 'This is Neovide!'
	en
endf

cal timer_start(2000, 'g:MyGuiInitCallback')
```

Is there any more or less normal why to detect Neovide at initialization step?

**Comment by clason** [comment, marked as answer]
`exists('g:neovide')`

**Reply by Kethku** [reply]
Note: I believe this is set before the users config is loaded.

The settings are set late because the system is waiting for the user's config to run in case they set initial values to something other than the defaults.

**Reply by unclechu** [reply]
Thanks! This seems to be exactly what I need, it is set already by initialization. I ended up with this universal (universal for my use-case) detection solution:
``` viml
if !exists('s:gui_app')
	let s:gui_app =
		\ { 'neovim_gtk': exists('g:GtkGuiLoaded')
		\ , 'neovide': exists('g:neovide')
		\ }
	for x in values(s:gui_app)
		if x | let s:gui_app.neovim_qt = 0 | en
	endfo
	if !has_key(s:gui_app, 'neovim_qt')
		let s:gui_app.neovim_qt = 1
	en
en
```

**Comment by clason** [comment]
Although neovide will read your `ginit.vim` and the standard `guifont`.

**Reply by unclechu** [reply]
I know. The point is that different GUI applications use my `ginit.vim` and different applications has different set up for the fonts. You can look at some mess I had to make in order to have a universal solution for both Neovim-Gtk and Neovim-Qt:

https://github.com/unclechu/neovimrc/blob/163278e150ce38220ab25ed0c5dead2ffdf31622/ginit.vim

Now I have to add support for Neovide too. And it’s way to manage GUI font settings is incompatible with neither of those two.

**Reply by clason** [reply]
I don't quite understand why you _have to_ add support for Neovide _in addition_ (personally, I'd stick with one GUI to make my life easier), but yes, that sounds tedious.
